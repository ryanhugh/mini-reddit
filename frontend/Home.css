.input {
    display: inline-block;
    max-width: 80%;
    margin-right: 12px;
    vertical-align: middle;
}

.submitButton {
	display: inline-block;
}

.imgContainer {
	height: 43px;
    max-width: 32px;
    display: inline-block;
}

.arrowUp {
	margin-bottom: -5px;
    margin-top: -5px;
}

.arrowDown {
	margin-bottom: -5px;
    margin-top: -5px;
}


.form {
    font-size: 24px;
}

.score {
	vertical-align: text-bottom; 
	font-size: 24px;
}

.text {
	vertical-align: text-bottom; 
	font-size: 24px;
	padding-left: 10px;
}

