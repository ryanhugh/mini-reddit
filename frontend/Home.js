import React from 'react';
import CSSModules from 'react-css-modules';
import { Panel, Navbar, FormControl, Button } from 'react-bootstrap';
import css from './Home.css';
import request from './request';

// These SVG files were taken from here: https://www.iconfinder.com/search/?q=arrow
import down from './down.svg';
import up from './up.svg';

// Home page component
class Home extends React.Component {
  constructor(props) {
    super(props);

    // Keep track of the state of this component.
    // Topics is the array of topics retrieved from the server.
    // Input is the current value of the text box.
    this.state = {
      topics: [],
      input: '',
    };

    // Bind methods so 'this' references the instance of Home
    this.onSubmit = this.onSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.upvote = this.upvote.bind(this);
    this.downvote = this.downvote.bind(this);

    // Start by downloading all the topics from the server
    this.fetchTopics();
  }


  // Verifies that the post if valid and then uploads it to the server.
  async onSubmit() {
    if (!this.state.text || this.state.text.length > 255) {
      return;
    }

    await request('/newPost', {
      text: this.state.text,
    });

    this.fetchTopics();
  }

  // Download all the topics from the server.
  async fetchTopics() {
    const topics = await request('/topics');

    this.setState({
      topics: topics,
    });
  }

  // Updates this.state.text with the value from the text box.
  handleChange(e) {
    this.setState({ text: e.target.value });
  }

  // Tells the server to upvote a post
  async upvote(id) {
    await request('/upvote', {
      id: id,
    });


    this.fetchTopics();
  }

  // Tells the server to downvote a post
  async downvote(id) {
    await request('/downvote', {
      id: id,
    });


    this.fetchTopics();
  }

  // Renders the page.
  render() {
    if (!this.state.topics) {
      return null;
    }

    // Sort the topics.
    const sortedTopics = this.state.topics.sort((a, b) => {
      if (a.score > b.score) {
        return -1;
      } else if (a.score < b.score) {
        return 1;
      }

      return 0;
    });


    return (
      <div>
        <Navbar>
          <Navbar.Header>
            <Navbar.Brand>
              <a href='/#'>Mini Reddit</a>
            </Navbar.Brand>
          </Navbar.Header>
        </Navbar>

        <div className='container bs-docs-container bs-docs-single-col-container'>

          <form className={ css.form }>
            <FormControl
              className={ css.input }
              type='text'
              placeholder='Enter text'
              onChange={ this.handleChange }
            />

            <Button onClick={ this.onSubmit } className={ css.submitButton }>
              Submit
            </Button>
          </form>

          {sortedTopics.map((topic) => {
            const upvote = this.upvote.bind(this, topic.id);
            const downvote = this.downvote.bind(this, topic.id);

            return (
              <Panel key={ topic.id }>
                <span className={ css.score }>
                  {topic.score}
                </span>

                <span className={ css.imgContainer }>
                  <img src={ up } className={ css.arrowUp } onClick={ upvote } alt='Upvote' />
                  <img src={ down } className={ css.arrowDown } onClick={ downvote } alt='Downvote' />
                </span>

                <span className={ css.text }>
                  {topic.text}
                </span>
              </Panel>
            );
          })}
        </div>
      </div>
    );
  }
}


export default CSSModules(Home, css);
