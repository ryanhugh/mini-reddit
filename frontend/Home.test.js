import React from 'react';

import renderer from 'react-test-renderer';

import Home from './Home';
jest.mock('./request');


it('should render a couple posts', () => {

  const tree = renderer.create(<Home></Home>)

  expect(tree).toMatchSnapshot();
});


it('can downvote a post', () => {

  const component = renderer.create(<Home></Home>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  component.getInstance().downvote();

  expect(component.toJSON()).toMatchSnapshot();
});


it('can upvote a post', () => {

  const component = renderer.create(<Home></Home>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  component.getInstance().upvote(1);

  expect(component.toJSON()).toMatchSnapshot();
});
