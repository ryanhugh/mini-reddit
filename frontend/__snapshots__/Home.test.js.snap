// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can downvote a post 1`] = `
<div>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container"
    >
      <div
        className="navbar-header"
      >
        <a
          className="navbar-brand"
          href="/#"
        >
          Mini Reddit
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container bs-docs-container bs-docs-single-col-container"
  >
    <form
      className="form"
    >
      <input
        className="input form-control"
        id={undefined}
        placeholder="Enter text"
        type="text"
      />
      <button
        className="submitButton btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`can downvote a post 2`] = `
<div>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container"
    >
      <div
        className="navbar-header"
      >
        <a
          className="navbar-brand"
          href="/#"
        >
          Mini Reddit
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container bs-docs-container bs-docs-single-col-container"
  >
    <form
      className="form"
    >
      <input
        className="input form-control"
        id={undefined}
        placeholder="Enter text"
        type="text"
      />
      <button
        className="submitButton btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`can upvote a post 1`] = `
<div>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container"
    >
      <div
        className="navbar-header"
      >
        <a
          className="navbar-brand"
          href="/#"
        >
          Mini Reddit
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container bs-docs-container bs-docs-single-col-container"
  >
    <form
      className="form"
    >
      <input
        className="input form-control"
        id={undefined}
        placeholder="Enter text"
        type="text"
      />
      <button
        className="submitButton btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`can upvote a post 2`] = `
<div>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container"
    >
      <div
        className="navbar-header"
      >
        <a
          className="navbar-brand"
          href="/#"
        >
          Mini Reddit
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container bs-docs-container bs-docs-single-col-container"
  >
    <form
      className="form"
    >
      <input
        className="input form-control"
        id={undefined}
        placeholder="Enter text"
        type="text"
      />
      <button
        className="submitButton btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`should render a couple posts 1`] = `
<div>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container"
    >
      <div
        className="navbar-header"
      >
        <a
          className="navbar-brand"
          href="/#"
        >
          Mini Reddit
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container bs-docs-container bs-docs-single-col-container"
  >
    <form
      className="form"
    >
      <input
        className="input form-control"
        id={undefined}
        placeholder="Enter text"
        type="text"
      />
      <button
        className="submitButton btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
